File: src/machine/SxxMachine/Real.cpp
using namespace std;

#include "Real.h"
#include "Prolog.h"
#include "Number.h"
#include "Appendable.h"

namespace SxxMachine {

	Real::Real(Number* i) {
		Num = i;
	}

	Term* Real::Copy(Prolog* m, long long t) {
		return new Real(Num);
	}

	long long Real::LongValue() {
		return Num->longValue();
	}

	Term* Real::Deref() {
		return this;
	}

	void Real::formattedOutput(const int& formatFlags, Appendable* buffer) throw(IOException) {
		buffer->append("" + Num);
	}

	bool Real::Unify(Term* that, Prolog* mach) {
		if(that->Arity() != Arity()) {
			return that->Bind(this, mach);
		}
		return (this->DoubleValue() == (this->DoubleValue()));
	}

	bool Real::Equal(Term* that) {
		if(that->Arity() != Arity()) {
			return false;
		}
		return (this->DoubleValue() == (this->DoubleValue()));
	}

	wstring Real::GetName() {
		return "" + Num;
	}

	int Real::Arity() {
		return Term::REAL;
	}

	double Real::DoubleValue() {
		return Num->doubleValue();
	}

	bool Real::isVar() {
		return false;
	}

	bool Real::isFVar() {
		return false;
	}

	bool Real::isStruct() {
		return false;
	}

	bool Real::isConst() {
		return false;
	}

	bool Real::isInt() {
		return Num->hashCode() == Num->intValue();
	}
}

File: jsrc/machine/SxxMachine/Real.java
package SxxMachine;

import java.io.IOException;;

public class Real extends Term {
	public Number Num;

	Real(Number i) {
		this.Num = i;
	}

	Term Copy(Prolog m, long t) {
		return new Real(this.Num);
	}

	long LongValue() {
		return this.Num.longValue();
	}

	public Term Deref() {
		return this;
	}

	public void formattedOutput(int formatFlags, Appendable buffer) throws IOException {
		buffer.append("" + this.Num);
	}

	public boolean Unify(Term that, Prolog mach) {
		if (that.Arity() != this.Arity()) {
			return that.Bind(this, mach);
		}
		return (this.DoubleValue() == (this.DoubleValue()));
	}

	boolean Equal(Term that) {
		if (that.Arity() != this.Arity()) {
			return false;
		}
		return (this.DoubleValue() == (this.DoubleValue()));
	}

	public String GetName() {
		return "" + this.Num;
	}

	public int Arity() {
		return Data.REAL;
	}

	public final double DoubleValue() {
		return this.Num.doubleValue();
	}

	@Override
	public boolean isVar() {
		return false;
	}

	@Override
	public boolean isFVar() {
		return false;
	}

	@Override
	public boolean isStruct() {
		return false;
	}

	@Override
	public boolean isConst() {
		return false;
	}

	@Override
	public boolean isInt() {
		return this.Num.hashCode() == this.Num.intValue();
	}

}